### ISSUE TITLE
2.14 900d22c6: s3 list returns at most 2000 results

### ISSUE THREAD

**Issue by ginkel** [chrislusf/seaweedfs#1656]
**Describe the bug**
Issuing a listing of the contents of a bucket stored in SeaweedFS via its S3 interface returns just 2000 hits instead of all files contained in the bucket. Steps to reproduce:

1. Upload 3.7 million files into a single bucket using rclone (via S3)
2. Issue `rclone ls seaweedfs:/bucket/ | wc -l`
3. Notice that only 2000 files are returned with the listing. `rclone sync` shows similar issues and will only sync 2000 files instead of all

**System Setup**

```
weed master -mdir=/data -volumePreallocate -volumeSizeLimitMB=1024 -ip=10.132.0.35 -mdir=/data -peers=10.132.0.35:9333,10.132.15.228:9333,10.132.0.39:9333 -defaultReplication=002 -volumeSizeLimitMB=256
weed volume -dir=/data -max=0 -ip=10.132.0.35 -mserver=10.132.0.35:9333,10.132.15.228:9333,10.132.0.39:9333 -dataCenter=dc -rack=rack -dir=/data0,/data1 -metricsPort=9080
weed filer -ip=10.132.0.35 -master=10.132.0.35:9333,10.132.15.228:9333,10.132.0.39:9333 -peers=10.132.0.35:8888,10.132.15.228:8888,10.132.0.39:8888,10.132.0.38:8888 -metricsPort=9088 -s3 -s3.config=/etc/seaweedfs/s3.json

# uname -a
Linux instance-seaweedfs-0 5.4.0-1030-gcp #32-Ubuntu SMP Fri Nov 13 12:13:03 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

# weed version
version 30GB 2.14 900d22c6 linux amd64
```

**Expected behavior**
The listing returns the number of files in the bucket.

**Screenshots**

**Additional context**

`filer.toml`:

```
[leveldb2]
enabled = true
dir = "/data"
```

`rclone.conf`:

```
[seaweedfs]
type = s3
provider = Other
env_auth = false
access_key_id = <access_key>
secret_access_key = <secret>
region =
location_constraint =
acl =
server_side_encryption =
storage_class =
endpoint = http://seaweedfs-int.<domain>
```

`s3.json`:

```
{
    "identities": [
        {
            "actions": [
                "Admin",
                "Read",
                "List",
                "Tagging",
                "Write"
            ],
            "credentials": [
                {
                    "accessKey": "<redacted>",
                    "secretKey": "<redacted>"
                }
            ],
            "name": "<redacted>"
        }
    ]
}
```

**Comment by chrislusf** [comment]
S3 has a limit of 1000. See https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html

Following S3 standard, there is a `maxObjectListSizeLimit` of 10000. There are actually nothing hard coded to 2000.

https://github.com/chrislusf/seaweedfs/blob/81024761cdb94ea75cf7e6917c026eb5cb84330c/weed/s3api/s3api_object_multipart_handlers.go#L17

**Comment by chrislusf** [comment]
btw: The limit can be dynamically changed via url parameter `max-keys`. Maybe `rclone` has adjusted the parameter?

**Comment by chrislusf** [comment]
can you verify your setting? https://rclone.org/s3/#s3-list-chunk

**Comment by ginkel** [comment]
I did a `tcpdump` to verify what's going on. The sequence apparently is:

```
GET /konserve?delimiter=&max-keys=1000&prefix= HTTP/1.1
Host: instance-seaweedfs-0-int....
User-Agent: rclone/v1.53.3
Authorization: AWS4-HMAC-SHA256 Credential=...
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20201208T225240Z
Accept-Encoding: gzip

HTTP/1.1 200 OK
Accept-Ranges: bytes
Content-Length: 272651
Content-Type: application/xml
Server: SeaweedFS S3 30GB 2.14
X-Amz-Request-Id: 1607467961021609924
Date: Tue, 08 Dec 2020 22:52:41 GMT

<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>konserve</Name><Prefix></Prefix><Marker></Marker><NextMarker>konserve/bahn-realtime-20201001/20201001003546_8000128.xml</NextMarker><MaxKeys>1000</MaxKeys><IsTruncated>true</IsTruncated><Contents>[...]

GET /konserve?delimiter=&marker=konserve%2Fbahn-realtime-20201001%2F20201001003546_8000128.xml&max-keys=1000&prefix= HTTP/1.1
Host: instance-seaweedfs-0-int....
User-Agent: rclone/v1.53.3
Authorization: AWS4-HMAC-SHA256 ...
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20201208T225241Z
Accept-Encoding: gzip

HTTP/1.1 200 OK
Accept-Ranges: bytes
Content-Length: 272539
Content-Type: application/xml
Server: SeaweedFS S3 30GB 2.14
X-Amz-Request-Id: 1607467961074353958
Date: Tue, 08 Dec 2020 22:52:41 GMT

<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>konserve</Name><Prefix></Prefix><Marker>konserve/bahn-realtime-20201001/20201001003546_8000128.xml</Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents>[...]
```

AFAICS `IsTruncated` is incorrectly set to `false` on the second page, isn't it?

**Comment by chrislusf** [comment]
The last response should have IsTruncated set to false. This seems to show there are only 2000 files in that directory.

My local env works fine. Can you try using this and see how many results are returned?

```
aws --endpoint-url http://localhost:8333 s3 ls s3://xxx/ --page-size=8000

```

**Comment by ginkel** [comment]
```
# aws --endpoint-url http://instance-seaweedfs-0:80 s3 ls s3://konserve/ --page-size=8000 --recursive | wc -l
16000

# aws --endpoint-url http://instance-seaweedfs-0:80 s3 ls s3://konserve/ --page-size=10000 --recursive | wc -l
20000

# aws --endpoint-url http://instance-seaweedfs-0:80 s3 ls s3://konserve/ --page-size=8000 --recursive | head -10
2020-12-09 19:21:50         65 konserve/bahn-realtime-20201001/20201001000000_8010060.xml
2020-12-09 19:21:50        384 konserve/bahn-realtime-20201001/20201001000002_8000181.xml
2020-12-09 19:21:50      14663 konserve/bahn-realtime-20201001/20201001000002_8002549.xml
2020-12-09 19:21:50         61 konserve/bahn-realtime-20201001/20201001000003_8010066.xml
2020-12-09 19:21:50       6415 konserve/bahn-realtime-20201001/20201001000003_8070003.xml
2020-12-09 19:21:50         61 konserve/bahn-realtime-20201001/20201001000005_8000182.xml
2020-12-09 19:21:50       6759 konserve/bahn-realtime-20201001/20201001000005_8002553.xml
2020-12-09 19:21:50        255 konserve/bahn-realtime-20201001/20201001000006_8010073.xml
2020-12-09 19:21:50       2850 konserve/bahn-realtime-20201001/20201001000007_8000013.xml
2020-12-09 19:21:50      27867 konserve/bahn-realtime-20201001/20201001000007_8000085.xml
```

My current hypothesis is that this is caused by the fact that this is triggered by the fact that I have a recursive directory structure in the `konserve` bucket.

**Comment by chrislusf** [comment]
Could you please make a reproducible case with your directory structure? It can be a small one and we can use`--page-size=3` to test.

**Comment by ginkel** [comment]
Sure. Is a tar file ok or do you need something more sophisticated?

**Comment by chrislusf** [comment]
a small tar file should be fine.

**Comment by ginkel** [comment]
There you go: https://github.com/ginkel/seaweed-s3-list-test

```
# aws --endpoint-url http://instance-seaweedfs-0:80 s3 ls s3://testbucket/ --page-size=3 --recursive
2020-12-09 23:24:20          1 dir/1/file001.bin
2020-12-09 23:24:20          1 dir/1/file002.bin
2020-12-09 23:24:20          1 dir/1/file003.bin
2020-12-09 23:24:20          1 dir/1/file004.bin
2020-12-09 23:24:20          1 dir/1/file005.bin
2020-12-09 23:24:20          1 dir/1/file006.bin

# aws --endpoint-url http://instance-seaweedfs-0:80 s3 ls s3://testbucket/ --page-size=1000 --recursive
2020-12-09 23:24:20          1 dir/1/file001.bin
2020-12-09 23:24:20          1 dir/1/file002.bin
2020-12-09 23:24:20          1 dir/1/file003.bin
2020-12-09 23:24:20          1 dir/1/file004.bin
2020-12-09 23:24:20          1 dir/1/file005.bin
2020-12-09 23:24:20          1 dir/1/file006.bin
2020-12-09 23:24:20          1 dir/1/file007.bin
2020-12-09 23:24:20          1 dir/1/file008.bin
2020-12-09 23:24:20          1 dir/1/file009.bin
2020-12-09 23:24:20          1 dir/1/file010.bin
2020-12-09 23:24:20          1 dir/2/file001.bin
2020-12-09 23:24:20          1 dir/2/file002.bin
2020-12-09 23:24:20          1 dir/2/file003.bin
2020-12-09 23:24:20          1 dir/2/file004.bin
2020-12-09 23:24:20          1 dir/2/file005.bin
2020-12-09 23:24:20          1 dir/2/file006.bin
2020-12-09 23:24:20          1 dir/2/file007.bin
2020-12-09 23:24:20          1 dir/2/file008.bin
2020-12-09 23:24:20          1 dir/2/file009.bin
2020-12-09 23:24:20          1 dir/2/file010.bin
2020-12-09 23:24:20          1 dir/3/file001.bin
2020-12-09 23:24:20          1 dir/3/file002.bin
2020-12-09 23:24:20          1 dir/3/file003.bin
2020-12-09 23:24:20          1 dir/3/file004.bin
2020-12-09 23:24:20          1 dir/3/file005.bin
2020-12-09 23:24:20          1 dir/3/file006.bin
2020-12-09 23:24:20          1 dir/3/file007.bin
2020-12-09 23:24:20          1 dir/3/file008.bin
2020-12-09 23:24:20          1 dir/3/file009.bin
2020-12-09 23:24:20          1 dir/3/file010.bin
```

Thanks!

**Comment by chrislusf** [comment]
cool! I can reproduce it now.

**Comment by chrislusf** [comment]
fixed. you can use the latest auto build https://github.com/chrislusf/seaweedfs/releases/tag/dev usually has a 4 minutes delay.

**Comment by ginkel** [comment]
LGTM, tested successfully. Thanks! :-)

**Comment by ginkel** [comment]
Just a follow-up question: After deploying the fixed version deleting the 3.7 mio files bucket is really slow (and consumes all available CPU while barely any I/O is happening). Could this be related to this fix?

Let me know if you'd rather have a separate issue for this. Thanks!
